pkg/rbac/sharedData.go

```diff
@@ -32,116 +39,171 @@ type resource struct {
+		if err == nil {
+			klog.V(6).Info("Successfully retrieved shared namespaces!")
+		} else {
+			error = err
+		}
+		// get all managed clustsers in cache
+		err = cache.shared.GetManagedClusters(cache, ctx)
+		if err == nil {
+			klog.V(6).Info("Successfully retrieved managed clusters!")
+		} else {
+			error = err
+		}
+		return error
+
+	}
+
+}
+
+func sharedCacheValid(shared *SharedData) bool {
+
+	if (time.Now().Before(shared.csUpdatedAt.Add(time.Duration(config.Cfg.SharedCacheTTL) * time.Millisecond))) &&
+		(time.Now().Before(shared.nsUpdatedAt.Add(time.Duration(config.Cfg.SharedCacheTTL) * time.Millisecond))) &&
+		(time.Now().Before(shared.mcUpdatedAt.Add(time.Duration(config.Cfg.SharedCacheTTL) * time.Millisecond))) {
+
+		return true
+	}
+	return false
+}
+
+// Obtain all the cluster-scoped resources in the hub cluster that support list and watch.
+// Equivalent to: `oc api-resources -o wide | grep false | grep watch | grep list`
+func (shared *SharedData) GetClusterScopedResources(cache *Cache, ctx context.Context) error {
 
 	// lock to prevent checking more than one at a time and check if cluster scoped resources already in cache
 	shared.csLock.Lock()
 	defer shared.csLock.Unlock()
-	if shared.csResources == nil ||
-		time.Now().After(shared.csUpdatedAt.Add(time.Duration(config.Cfg.SharedCacheTTL)*time.Millisecond)) {
-
-		// if not in cache query database
-		klog.V(6).Info("Querying database for cluster-scoped resources.")
-		shared.csErr = nil // Clear previous errors.
-
-		// Building query to get cluster scoped resources
-		// Original query: "SELECT DISTINCT(data->>apigroup, data->>kind) FROM search.resources WHERE
-		// cluster='local-cluster' AND namespace=NULL"
-
-		schemaTable := goqu.S("search").Table("resources")
-		ds := goqu.From(schemaTable)
-		query, _, err := ds.SelectDistinct(goqu.COALESCE(goqu.L(`"data"->>'apigroup'`), "").As("apigroup"),
-			goqu.COALESCE(goqu.L(`"data"->>'kind_plural'`), "").As("kind")).
-			Where(goqu.L(`"cluster"::TEXT = 'local-cluster'`), goqu.L(`"data"->>'namespace'`).IsNull()).ToSQL()
-		if err != nil {
-			klog.Errorf("Error creating query [%s]. Error: [%+v]", query, err)
-			shared.csErr = err
-			shared.csResources = []resource{}
-			return shared.csResources, shared.csErr
-		}
+	//clear previous cache
+	shared.csResources = nil
+	shared.csErr = nil
+	klog.V(6).Info("Querying database for cluster-scoped resources.")
+
+	// Building query to get cluster scoped resources
+	// Original query: "SELECT DISTINCT(data->>apigroup, data->>kind) FROM search.resources WHERE
+	// cluster='local-cluster' AND namespace=NULL"
+	schemaTable := goqu.S("search").Table("resources")
+	ds := goqu.From(schemaTable)
+	query, _, err := ds.SelectDistinct(goqu.COALESCE(goqu.L(`"data"->>'apigroup'`), "").As("apigroup"),
+		goqu.COALESCE(goqu.L(`"data"->>'kind_plural'`), "").As("kind")).
+		Where(goqu.L(`"cluster"::TEXT = 'local-cluster'`), goqu.L(`"data"->>'namespace'`).IsNull()).ToSQL()
```

Comment:
From our discussion yesterday, use data->>'_hubClusterResource'='true' instead of "cluster"::TEXT = 'local-cluster' to avoid the usage of `local-cluster`